### ISSUE TITLE
system prompt not supported on gemma models

### ISSUE THREAD

**Issue by skunkwerk** [stanfordnlp/dspy#1627]
I'm running the [minimal example](https://dspy-docs.vercel.app/docs/quick-start/minimal-example) with the google/gemma-2-9b-it model.

Running with vLLM's OpenAI server, I get this error:

```litellm.exceptions.BadRequestError: litellm.BadRequestError: OpenAIException - Error code: 400 - {'object': 'error', 'message': 'System role not supported', 'type': 'BadRequestError', 'param': None, 'code': 400}```

Is there a way to run DSPy with models that don't support system prompts?

thanks

**Comment by arnavsinghvi11** [comment]
Hi @skunkwerk , does this discussion from this [LiteLLM issue](https://github.com/BerriAI/litellm/issues/2831#issuecomment-2074284723) help?

**Comment by okhat** [comment]
Hey @skunkwerk ! You'd have to build an adapter.

You can simply build an adapter that calls ChatAdapter and just replaces "system" roles with "user" roles. It could be done in 5-10 lines in total.

See the migration notebook for info about adapters: https://github.com/stanfordnlp/dspy/blob/main/examples/migration.ipynb

See ChatAdapter implementation: https://github.com/stanfordnlp/dspy/blob/main/dspy/adapters/chat_adapter.py

**Comment by skunkwerk** [comment]
thanks Prof. @okhat!  Here's what I'm going to test out:

```
import dspy

class GemmaChatAdapter(dspy.ChatAdapter):
    def format(self, signature, demos, inputs):
        messages = super().format(signature, demos, inputs)
        if messages and messages[0]["role"] == "system":
            messages[0]["role"] = "user"
        return messages

dspy.configure(adapter=GemmaChatAdapter())
```

**Comment by okhat** [comment]
Nice, @skunkwerk ! Lmk if it works!

**Comment by skunkwerk** [comment]
Yeah, it worked fine!  But then ran into [this issue](https://github.com/stanfordnlp/dspy/issues/1132), as vLLM only supports 4k context length with gemma :(. So now trying it out with TGI's OpenAI server

**Comment by windyeh** [comment]
due to the `jinja2.exceptions.TemplateError: Conversation roles must alternate user/assistant/user/assistant/...`
workaround by adding a dummy message.

```python
import dspy

class GemmaChatAdapter(dspy.ChatAdapter):
    def format(self, signature, demos, inputs):
        messages = super().format(signature, demos, inputs)
        if messages and messages[0]["role"] == "system":
            messages[0]["role"] = "user"
            messages.insert(1, {"role": "assistant", "content": "noted"})
        return messages

dspy.configure(adapter=GemmaChatAdapter())

```
